Use miliseconds for statistics

bendavies opened the pull request:
This is a rebased submission of #27 

fixes https://github.com/SymfonyCasts/messenger-monitor-bundle/issues/20
fixes https://github.com/SymfonyCasts/messenger-monitor-bundle/issues/17
fixes https://github.com/SymfonyCasts/messenger-monitor-bundle/issues/25

TODO: 

- [ ] document data migration

nikophil committed:
Use miliseconds for statistics

store waiting_time in seconds+microseconds

store handling_time in seconds+microseconds

store failing_time in seconds+microseconds

Fixed statistics computing

Handle micro/milli seconds in twig time filter

Tested retry behavior

Fixed waiting_time for messages with delay

weaverryan commented on src/Storage/Doctrine/Connection.php:
```diff
@@ -181,10 +175,10 @@ private function addTableToSchema(Schema $schema): void
+        $table->addColumn('dispatched_at', Types::FLOAT)->setNotnull(true);
+        $table->addColumn('waiting_time', Types::FLOAT)->setNotnull(false);
+        $table->addColumn('handling_time', Types::FLOAT)->setNotnull(false);
+        $table->addColumn('failing_time', Types::FLOAT)->setNotnull(false);
```
Just an FYI for us, this will require a migration - iirc, `make:migration` will see this changes, but it's been awhile so I'm not 100% sure.

bendavies replied:
looks like you can backfill these new columns, as they are calculated from the existing ones?
the existing ones can then be dropped

weaverryan reviewed (dismissed):
There's quite a bit going on here, but i'm 👍  and the tests look solid.

weaverryan commented:
@bendavies now that CI is running, there's one tiny failure related to precision. Can you have a look?

bendavies commented:
Interesting. That didn't happen on the original pr

bendavies commented:
> @bendavies now that CI is running, there's one tiny failure related to precision. Can you have a look?

looks like this is due to floating point calculations in mariadb
<img width="508" alt="image" src="https://user-images.githubusercontent.com/625392/193228453-32e6d7ce-168c-4325-9fd6-9abe3922aa54.png">


i'll see if switching the column types from float to decimal fixes this.

bendavies committed:
store times as numerics, not floats

floats are inexact: https://www.postgresql.org/docs/current/datatype-numeric.html#DATATYPE-FLOAT
numeric are exact: https://www.postgresql.org/docs/current/datatype-numeric.html#DATATYPE-NUMERIC-DECIMAL

bendavies commented:
using decimals has fixed the tests
